[ios] [android] Refactor the visibility and enabling states of the Add/Edit Place buttons on the PP and Menu

kirylkaveryn opened the pull request:
### Issue description:
The `Add Place to OSM` can be selected even in disabled (text is gray) mode in ios.

### Expected behavior:
1. The `Add/Edit place` on the PP should be disabled when the map is not downloaded OR outdated (>3mo)
2. The `Add place` on the Menu should be disabled when the map is not downloaded OR outdated (>3mo)
3. If the user open the `Add place` from the `Menu` and moves the map an additional check should happens to prevent editing of the outdated map
4. The `Add/Edit place` on the PP should be visible even if disabled (as a next step it will be possible to change the title to the "Update/Download [name of the region]" and navigate the user to the downloading page from this button).

## This PR contains:
### Core
1. The logic to check if the map/poi is editable is moved from the `mwm_info` to the `storage` and based on the file's `state` and `version` because this check should be used for both 1) the selected poi editing and 2) the add place from the menu only by using only the viewport coordinates (to deactivate the `add place` when the `viewport center` is out of the downloaded OR updated area).
2. `CanEditMap()` changed to the `CanEditMapForPosition(m2::PointD const & position)` because the previous method only checks the downloading of some map progress - this is not correct when the user tries to edit the fresh map while some other region is downloaded:
```
bool Framework::CanEditMap() const
{
  return !GetStorage().IsDownloadInProgress();
}
```
3. `bool ShouldEnableAddPlace() const { return m_canEditOrAdd; };` and `bool ShouldEnableEditPlace() const { return m_canEditOrAdd; };` were added to separate the add/edit buttons editable state from the visibility.

@vng can you please review the changes?

### iOS
1. Menu screen:
1.1. If the current viewport center points into the **downloaded latest or outdated (< that 3 months)** map the `Add Place` button will be **enabled**
1.2. If the current viewport center points into the **not downloaded or outdated (> that 3 months)** map the `Add Place` button will be **disabled**

3. Place Page screen
2.1. If the selected POI is on the **downloaded latest or outdated (< that 3 months)** map the `Add Place/Edit Place` buttons will be visible and **enabled**
2.2. If the selected POI is on the **not downloaded or outdated (> that 3 months)** map the `Add Place/Edit Place` buttons will be visible and **disabled**

#### Downloaded latest or outdated (< that 3 months):
<img width="350" alt="image" src="https://github.com/user-attachments/assets/17f8b965-be0a-4285-bb1f-5c6cee51ffcb">
<img width="350" alt="image" src="https://github.com/user-attachments/assets/a1533e80-c705-4d68-baea-016650ad8b9e">
<img width="350" alt="image" src="https://github.com/user-attachments/assets/fc89901d-8fdd-4a24-8554-b3e2f74fb38a">

#### Not downloaded maps or outdated (> that 3 months):
https://github.com/user-attachments/assets/fa667fe6-f511-4c15-99c3-460590f0c2a5
<img width="350" alt="image" src="https://github.com/user-attachments/assets/188138f1-4ee6-4ad8-a719-e5dc6d7d3584">
<img width="350" alt="image" src="https://github.com/user-attachments/assets/4d829e52-460b-4d6e-942b-5e188f9fa76a">

Menu:

![Simulator Screen Recording - iPhone 15 - 2024-07-23 at 19 48 52](https://github.com/user-attachments/assets/437296c6-cc0d-4bbb-875a-cb477894a0e1)
![Simulator Screen Recording - iPhone 15 - 2024-07-23 at 19 49 40](https://github.com/user-attachments/assets/1d5accfb-c19c-48f1-ad56-1fac1155cf20)

### Android:
1. Place Page screen
1.1. If the selected POI is on the **downloaded latest or outdated (< that 3 months)** map the `Add Place/Edit Place` buttons will be visible and **enabled**
1.2. If the selected POI is on the **not downloaded or outdated (> that 3 months)** map the `Add Place/Edit Place` buttons will be visible and **disabled**

TODO:
- [ ] disabled state for the PP's buttons on the Android should be grayed (not just not active)

#### Downloaded latest maps:
<img width="350" alt="image" src="https://github.com/user-attachments/assets/9edca734-aa1c-4781-8409-d6755abbef9f">

#### Not downloaded maps:
<img width="350" alt="image" src="https://github.com/user-attachments/assets/4dd05454-9f2e-4a97-b710-46aa298e1b1a">


### Key differences from the previous implementation:
- there are no situations when both of the `Add Place/Edit Place` buttons are hidden (it happens when map downloading is in progress regardless of where the user tries to add a place), so they will be disabled (in the future they should contain text to notify a user that is needed to download/update the map).
- maps that are outdated but not too much (< that 3 months) can be edited

biodranik commented:
There are two issues here:
1. Greyed cells should not be selected.
2. The cell should not be greyed if the underlying map region is updated (there can be a double-check: check if the center of the displayed map is updated when opening a menu, and then check if the center of the map - crosshair - is updated, because a user can move the crosshair)

kirylkaveryn commented:
@biodranik 
Updated behavior according to your notice: 
- if the map is not downloaded / outdated the Add/Edit place will be fully blocked
- the check will be done twice: on the Menu opening and when the Done button is pressed

Results:

![Simulator Screen Recording - iPhone 15 - 2024-07-23 at 19 48 52](https://github.com/user-attachments/assets/437296c6-cc0d-4bbb-875a-cb477894a0e1)
![Simulator Screen Recording - iPhone 15 - 2024-07-23 at 19 49 40](https://github.com/user-attachments/assets/1d5accfb-c19c-48f1-ad56-1fac1155cf20)

Downloaded and Updated map:
<img width="350" alt="image" src="https://github.com/user-attachments/assets/0abf4559-88e7-4485-bf99-a72f8df1d705">

Not Downloaded map:
<img width="350" alt="image" src="https://github.com/user-attachments/assets/abf6e074-e8ab-45f4-8ded-d68c253ea15f">

Downloaded and NOT Updated map:
<img width="350" alt="image" src="https://github.com/user-attachments/assets/f5e23ffd-de3f-4371-92cc-2c8b0a217845">
<img width="350" alt="image" src="https://github.com/user-attachments/assets/ca5027e0-cf14-4adc-af3c-5d5da09af2c9">

biodranik commented:
@kirylkaveryn it would be great to also add "Update outdated maps" button somewhere, either on the Place Page or at the bottom bar, or in the editor screen (when editing and adding a feature).

biodranik commented:
@kirylkaveryn does it make sense to change "Add..." and "Edit..." buttons into "Update [name of the region]"?

kirylkaveryn commented:
@biodranik it makes sense! I'll make them tappable in ios in the next RP (as we discussed earlier and based on the design https://www.figma.com/design/zxnZsa3SBaSAPWG0BmZ3Xh/OM-UpdateMap?node-id=0-1&t=0hx1uFqPvf2DBjIw-1).

kirylkaveryn commented:
@Jean-BaptisteC can you please help to test the PR on the Android (see description) and fix the disabled state for the Add/Edit Place button on the PP screen (they should be grayed while disabled but not blue)?

Jean-BaptisteC commented:
You can try this
```
TextView mTvEditPlace = mEditPlace.findViewById(R.id.tv__editor);
mTvEditPlace.setTextColor(Editor.nativeShouldEnableEditPlace() ? getResources().getColor(R.color.base_accent) : getResources().getColor(R.color.bg_menu));
TextView mTvAddPlace = mAddPlace.findViewById(R.id.tv__editor);
mTvAddPlace.setTextColor(Editor.nativeShouldEnableEditPlace() ? getResources().getColor(R.color.base_accent) : getResources().getColor(R.color.bg_menu));
```

kirylkaveryn commented:
> You can try this
> 
> ```
> TextView mTvEditPlace = mEditPlace.findViewById(R.id.tv__editor);
> mTvEditPlace.setTextColor(Editor.nativeShouldEnableEditPlace() ? getResources().getColor(R.color.base_accent) : getResources().getColor(R.color.bg_menu));
> TextView mTvAddPlace = mAddPlace.findViewById(R.id.tv__editor);
> mTvAddPlace.setTextColor(Editor.nativeShouldEnableEditPlace() ? getResources().getColor(R.color.base_accent) : getResources().getColor(R.color.bg_menu));
> ```

@Jean-BaptisteC thanks!
I've updated the disabled color to the button_accent_text_disabled because the bg_menu is too light:

bg_menu
<img width="300" alt="image" src="https://github.com/user-attachments/assets/9d8be450-c0c2-4c98-bd45-a8440bb6a6fa">

button_accent_text_disabled
<img width="300" alt="image" src="https://github.com/user-attachments/assets/538156a3-6f99-46ff-9dac-6aaf2d2ff001">

vng commented on map/framework.cpp:
```diff
@@ -2855,9 +2854,11 @@ void SetHostingBuildingAddress(FeatureID const & hostingBuildingFid, DataSource
-  return !GetStorage().IsDownloadInProgress();
+  auto const & storage = GetStorage();
+  ASSERT(m_infoGetter, ("CountryInfoGetter shouldn't be nullprt."));
+  return IsPointCoveredByDownloadedMaps(position, storage, *m_infoGetter) && storage.IsAllowedToEditVersion(m_infoGetter->GetRegionCountryId(position));
```
I suspect that it should work without first IsPointCoveredByDownloadedMaps:
GetRegionCountryId -> returns empty -> status returns NotDownloaded.

kirylkaveryn replied:
Fixed and tested - it works as expected without the `IsPointCoveredByDownloadedMaps`. Thanks!

biodranik commented on iphone/CoreApi/CoreApi/Framework/MWMFrameworkHelper.mm:
```diff
@@ -162,8 +162,9 @@ + (void)searchInDownloader:(NSString *)query
-+ (BOOL)canEditMap {
-  return GetFramework().CanEditMap();
++ (BOOL)canEditMapAtViewportCenter {
+  auto &f = GetFramework();
```
```suggestion
  auto const & f = GetFramework();
```

kirylkaveryn replied:
fixed

biodranik commented on storage/storage.cpp:
```diff
@@ -1269,6 +1278,22 @@ bool Storage::HasLatestVersion(CountryId const & countryId) const
+    {
+      auto const localFile = GetLatestLocalFile(countryId);
+      ASSERT(localFile, ("Local file shouldn't be nullptr."));
+      return m_currentVersion - localFile->GetVersion() < kOutdatedVersionDiffLimit;
```
How does it work? For example `250101 - 241212 = 8889` but the difference is less than a month between these dates.

kirylkaveryn replied:
Thanks! Replace this check with the timeInterval.
Also, I added a check to compare the current map version with the current time (now) to avoid situations where devices without a disconnected from the internet for a long time can update the map (6mo).

biodranik commented on storage/storage.cpp:
```diff
@@ -42,6 +43,16 @@ namespace
+// is most likely already fixed on OSM. Not limited to the latest one or two versions,
+// because a user can forget to update maps after a new app version has been installed
+// automatically in the background.
+uint64_t const kMaxSecondsTillLastVersionUpdate = 3600 * 24 * 31 * 3;
```
```suggestion
uint64_t constexpr kMaxSecondsTillLastVersionUpdate = 3600 * 24 * 31 * 3;
```

biodranik commented on storage/storage.cpp:
```diff
@@ -42,6 +43,16 @@ namespace
+uint64_t const kMaxSecondsTillLastVersionUpdate = 3600 * 24 * 31 * 3;
+// Editing maps older than approximately six months old is disabled, because the device
+// may have been offline for a long time.
+uint64_t const kMaxSecondsTillNoEdits = 3600 * 24 * 31 * 6;
```
```suggestion
uint64_t constexpr kMaxSecondsTillNoEdits = 3600 * 24 * 31 * 6;
```

biodranik commented on storage/storage.cpp:
```diff
@@ -1269,6 +1280,25 @@ bool Storage::HasLatestVersion(CountryId const & countryId) const
+  auto const status = CountryStatusEx(countryId);
+  switch (status) 
+  {
+    case Status::OnDisk: return true;
```
If editing should be disabled in case of a very old, not updated app version, then 6 months period check should also be present here.

biodranik commented on storage/storage.cpp:
```diff
@@ -1269,6 +1280,25 @@ bool Storage::HasLatestVersion(CountryId const & countryId) const
+    case Status::OnDiskOutOfDate:
+    {
+      auto const localFile = GetLatestLocalFile(countryId);
+      ASSERT(localFile, ("Local file shouldn't be nullptr."));
```
nit: Some obvious checks don't require any comment ) like `ASSERT(ptr, ());`

biodranik commented on storage/storage.cpp:
```diff
@@ -1269,6 +1280,25 @@ bool Storage::HasLatestVersion(CountryId const & countryId) const
+      auto const localFile = GetLatestLocalFile(countryId);
+      ASSERT(localFile, ("Local file shouldn't be nullptr."));
+      auto const currentVersionTime = base::YYMMDDToSecondsSinceEpoch(static_cast<uint32_t>(m_currentVersion));
+      auto const localVersionTime = base::YYMMDDToSecondsSinceEpoch(static_cast<uint32_t>(localFile->GetVersion()));
```
outdatedVersionTime would be clearer, as both newer and older maps are local.

biodranik reviewed (approved):
Thanks, LGTM

@Jean-BaptisteC can you please help with testing it on Android? Maybe it would be possible to install some > 6 months old version and test it too with updated logic?

@vng PTAL and merge if it's ok

biodranik commented:
@Jean-BaptisteC that's strange, which fdroid version did you try?

Jean-BaptisteC commented:
Here https://t.me/OrganicMapsApp/388?single

biodranik commented:
These are betas with unreleased temporary data files, so it's ok. Please try official releases from github (you may need to uninstall the main app first for these older versions, backup your bookmarks!) https://github.com/organicmaps/organicmaps/releases/tag/2023.12.20-4-android

Jean-BaptisteC reviewed (commented):
On Android, buttons are disabled only during users download update maps. I have a version with map from June and I can edit map

biodranik commented:
So it works, thanks a lot to everyone!
